Subject: [PATCH] gitweb: allow basic refname suffixes

When using gitweb, sometimes the hash or ref of an object is not
readily available, but that of one of its children is.

Allow the use of ~n or ^n suffixes on refnames to facilitate using
child hash or ref names (e.g. child^ or child~2) in place of the
parent hash or ref name.

This can be particularly useful in following blame trails.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>
---
 gitweb/gitweb.perl | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/gitweb/gitweb.perl b/gitweb/gitweb.perl
index 8e2fbefa..2e8c41c4 100755
--- a/gitweb/gitweb.perl
+++ b/gitweb/gitweb.perl
@@ -1854,6 +1854,8 @@ sub is_valid_refname {
 	if ($input =~ m/^[0-9a-fA-F]{40}$/) {
 		return 1;
 	}
+	# allow repeated trailing '[~^]n*' suffix(es)
+	$input =~ s/^([^~^]+)(?:[~^]\d*)+$/$1/;
 	# it must be correct pathname
 	is_valid_pathname($input) or return undef;
 	# check git-check-ref-format restrictions
---
